### PR TITLE
Bluebird ES6 compatibility

### LINKED ISSUE #4112: Bluebird definitions conflict with lib.es6.d.ts
The bluebird definitions attempt to declare a global Promise class.  This conflicts with TypeScript's definitions for ES6's built-in promises, causing compiler errors.

https://github.com/Microsoft/TypeScript/blob/master/bin/lib.es6.d.ts#L4819

https://github.com/borisyankov/DefinitelyTyped/blob/master/bluebird/bluebird.d.ts#L19

### PR DESCRIPTION
**Pull request by cspotcode**
_This PR is a work-in-progress._  
Fixes #4112.  Duplicate of PR #4788, but the latter hasn't been touched in a while.

To properly support ES6, this type definition needs to be split into a couple declarations to support the different scenarios in which it might be used.
#### To-Do
- [ ] Split `BluebirdPromise` class into two interfaces: `BluebirdPromise` and `BluebirdPromiseConstructor`
  - This allows us to extend the built-in `Promise` and `PromiseConstructor` interfaces from `lib.es6.d.ts`
- [ ] `declare module "bluebird" {}`
  - allows us to `require('bluebird')`
- [ ] `declare module Promise {}`
  - exposes all Bluebird interfaces on the global `Promise` variable
#### Questions
- [ ] Do we need to write a declaration that will `declare var Promise: BluebirdPromiseConstructor;`?
  - In other words, will our declaration ever be used without `lib.es6.d.ts`?  
    Or does `tsc` _always_ load `lib.es6.d.ts`?
#### Automatically generating multiple versions of the declaration

Ideally, we'd have several bluebird declarations:
- "bluebird.d.ts": declares external "bluebird" module
- "bluebird-global.d.ts": extends global `Promise` variable _and_ declares external "bluebird" module

Due to TS quirks, the latter can't merely reference the former.

JS templating systems are a dime-a-dozen.  Devise a simple solution for building multiple versions of the declarations with a build script.  Something like C++ `#ifdef` proprocessor.  Make sure the source code is still valid .d.ts so that IDE Intellisense still works while writing declarations.

### CODE REVIEW

**Comment by vojtechhabarta** [comment]
Is it really necessary to generate multiple versions?
In #4788 comments was mentioned some proposed solution: http://goo.gl/bcz8Gl which should work well in both environments - browser and node.

**Comment by cspotcode** [comment]
@vojtechhabarta: Last I checked, the proposed solution will not work because this line throws an error:
```
declare var Promise: typeof bluebird.Promise;
```
`Promise` is already declared in `lib.es6.d.ts`.

**Comment by vojtechhabarta** [comment]
@cspotcode aha ok, and what about this?
```TypeScript

// bluebird.d.ts
interface Promise<T> {
    then<TResult>(onfulfilled?: (value: T) => TResult | PromiseLike<TResult>, onrejected?: (reason: any) => TResult | PromiseLike<TResult>): Promise<TResult>;  // ES6
	done: any;  // bluebird
}
interface PromiseConstructor {
    resolve(): Promise<void>;  // ES6
    cast: any;  // bluebird
}

// bluebird-browser.d.ts 
declare var Promise: PromiseConstructor;

// bluebird-node.d.ts
declare module "bluebird" {
    var p: PromiseConstructor;
    export = p;
}

```

**Comment by cspotcode** [comment]
That works if you don't mind extending the global Promise interface, and it's appropriate if bluebird is replacing the global `Promise` object.  However, if you're only using bluebird as a CommonJS module, then the built-in Promise interface shouldn't have any bluebird-specific methods on it.  The global `Promise` isn't the same as `require('bluebird')`.

**Comment by vojtechhabarta** [comment]
Thanks for your explanation. So if I understand it correctly adding bluebird functions to `Promise` and `PromiseConstructor` interfaces defined in `lib.es6.d.ts` means:
1. browser scenario is correct
2. `import * as Promise from "bluebird"` in node works (Promise import hides build in Promise)
3. import as something else is **incorrect**: imported bluebird module has appropriate functions but Promise type suddenly has all bluebird extra functions that ES6 Promise object doesn't have.

I think this is still better than current state when compilation fails but I agree that it is not perfect solution.

Thank you again

**Comment by cspotcode** [comment]
@vojtechhabarta Agreed, it would be better than the current state but not perfect.  And both solutions -- the perfect and the imperfect -- require splitting the current definition into two interfaces.  Whoever does the split, it'll be time well spent.

**Comment by mhegazy** [comment]
Thank you for your contribution. 

Can you please fix the failing Travis CI tests and update the PR!

**Comment by mhegazy** [comment]
Thanks for your contribution. This PR has failing CI tests and can not be merged in at the time being. For housekeeping purposes we are closing stale PRs. If you'd still like to merge this code in, please open a new PR that has been merged and rebased with the `master` branch.
